### PR TITLE
Add documentation for operators and subscripting

### PR DESCRIPTION
**Pull request by chamons**

### CODE REVIEW

**Comment by spouliot** [comment]
c.c. @timrisi for another pair of eyes

File: docs/BestPracticesObjC.md [hunk 1]
```diff
@@ -125,3 +125,11 @@ In .NET a constructor must either succeed and return a (_hopefully_) valid insta
 In contrast, ObjC allows `init*` to return `nil` when an instance cannot be created. This is a common, but not general, pattern used in many of Apple's frameworks. In some other cases an `assert` can happen (and kill the current process).
 
 The generator follow the same `return nil` pattern for generated `init*` methods. If a managed exception is thrown, then it will be printed (using `NSLog`) and `nil` will be returned to the caller.
+
+## Operators
+
+ObjC does not allow operators to be overloaded as C# does and these are converted to class selectors.
```
Comment (timrisi) [inline docs/BestPracticesObjC.md]: Maybe:

ObjC does not allow operators to be overloaded as C# does, so these are converted to class selectors.

File: docs/BestPracticesObjC.md [hunk 2]
```diff
@@ -125,3 +125,11 @@ In .NET a constructor must either succeed and return a (_hopefully_) valid insta
...
 
 The generator follow the same `return nil` pattern for generated `init*` methods. If a managed exception is thrown, then it will be printed (using `NSLog`) and `nil` will be returned to the caller.
+
+## Operators
+
+ObjC does not allow operators to be overloaded as C# does and these are converted to class selectors. 
+
+["Friendly"](https://msdn.microsoft.com/en-us/library/ms229032(v=vs.110).aspx) named method are generated in preference to the operator overloads when found and can produce an easier to consume API.
+
+Classes that override operator == and\or != should override the standard Equals (Object) method as well.
```
Comment (timrisi) [inline docs/BestPracticesObjC.md]: Classes that override the operators == and\or != should override the standard Equals (Object) method as well.

File: docs/BestPracticesObjC.md [hunk 3]
```diff
@@ -125,3 +125,11 @@ In .NET a constructor must either succeed and return a (_hopefully_) valid insta
 In contrast, ObjC allows `init*` to return `nil` when an instance cannot be created. This is a common, but not general, pattern used in many of Apple's frameworks. In some other cases an `assert` can happen (and kill the current process).
 
 The generator follow the same `return nil` pattern for generated `init*` methods. If a managed exception is thrown, then it will be printed (using `NSLog`) and `nil` will be returned to the caller.
+
+## Operators
+
+ObjC does not allow operators to be overloaded as C# does and these are converted to class selectors. 
+
+["Friendly"](https://msdn.microsoft.com/en-us/library/ms229032(v=vs.110).aspx) named method are generated in preference to the operator overloads when found and can produce an easier to consume API.
```
Comment (timrisi) [inline docs/BestPracticesObjC.md]: ["Friendly"](https://msdn.microsoft.com/en-us/library/ms229032(v=vs.110).aspx) named method are generated in preference to the operator overloads when found, and can produce an easier to consume API.

(Comma)

File: docs/ObjC.md [hunk 4]
```diff
@@ -106,4 +134,49 @@ Objective-C surfaced API:
...
+	}
+```
+
+to
+
+```
++ (instancetype)addition:(Overloads_AllOperators *)anObjectC1 c2:(Overloads_AllOperators *)anObjectC2;
+```
+
+However, some .NET languages do not support operator overloading so it is common to also include a ["friendly"](https://msdn.microsoft.com/en-us/library/ms229032(v=vs.110).aspx) named method in addition.
```
Comment (timrisi) [inline docs/ObjC.md]: However, some .NET languages do not support operator overloading, so it is common to also include a ["friendly"](https://msdn.microsoft.com/en-us/library/ms229032(v=vs.110).aspx) named method in addition to the operator overload.

(Comma + extra)

File: docs/ObjC.md [hunk 5]
```diff
@@ -106,4 +134,49 @@ Objective-C surfaced API:
...
+
+becomes:
+
+```
++ (instancetype)addC1:(Overloads_AllOperatorsWithFriendly *)anObjectC1 c2:(Overloads_AllOperatorsWithFriendly *)anObjectC2;
+```
+
+### Equality operator
+
+In general operator == in C# is handled as a general operators as noted above.
```
Comment (timrisi) [inline docs/ObjC.md]: In general operator == in C# is handled as a general operator as noted above.

(No plural on operator)

File: docs/ObjC.md [hunk 6]
```diff
@@ -106,4 +134,49 @@ Objective-C surfaced API:
...
+	public static AllOperatorsWithFriendly Add (AllOperatorsWithFriendly c1, AllOperatorsWithFriendly c2)
+	{
+		return new AllOperatorsWithFriendly (c1.Value + c2.Value);
+	}
+```
+
+becomes:
+
+```
++ (instancetype)addC1:(Overloads_AllOperatorsWithFriendly *)anObjectC1 c2:(Overloads_AllOperatorsWithFriendly *)anObjectC2;
```
Comment (timrisi) [inline docs/ObjC.md]: So out of curiosity, why do we use the "friendly" names if friendly versions are found, but not default to the friendly terms for the objc overload of an operator?  E.g. here we have the selector ```addC1:c2```, while earlier with just the + overload, we give the selector ```addition:c2:```  Why don't we use the "friendly" names (Add, Subtract, Multiply, Divide) for the class selector if they only override the operator?

File: docs/ObjC.md [hunk 7]
```diff
@@ -106,4 +134,49 @@ Objective-C surfaced API:
...
+	public static AllOperators operator + (AllOperators c1, AllOperators c2)
+	{
+		return new AllOperators (c1.Value + c2.Value);
+	}
+```
+
+to
+
+```
++ (instancetype)addition:(Overloads_AllOperators *)anObjectC1 c2:(Overloads_AllOperators *)anObjectC2;
```
Comment (timrisi) [inline docs/ObjC.md]: Shouldn't the selector be: ```additionC1:c2```?  Missing argument name on the first argument

Comment (chamons) [inline docs/ObjC.md]: I copy pasted that from generated code. That's what we're spitting out.

**Comment by chamons** [comment]
❤️  the doc corrections. FIxing.

File: docs/ObjC.md (hunk 6 above)
Comment (chamons) [inline docs/ObjC.md]: https://github.com/mono/Embeddinator-4000/issues/204

In theory they should generate the same name (and just go to a different C# method).

**Pull request merged**